### PR TITLE
Password modification and password policy features for user_ldap

### PR DESCRIPTION
**Pull request by GitHubUser4234**
_This is an enhancement I'm currently developing with mentorship from @blizzz :+1: , any suggestions for improvements or contributions are welcome._

**Ability to change LDAP password**

With the new features added to user_ldap in #23992, an app would be able to change a password in LDAP using a `pre_setPassword()` hook, when an ownCloud admin is doing it. 

However there are issues:
1. `setPassword()` is not implemented by user_ldap. Therefore always false is returned to the Controller and a message "Unable to change password" appears, even though the password change was successful in the app's `pre_setPassword()` hook
2. user's cannot change their own password as `setPassword()` is not implemented

Both issues could be solved by implementing @blizzz 's suggestions mentioned in #3785 instead of using an app. So let's assume I'll do it, which place is most suitable in the UI for the checkBox for setting the "enabled" flag in the `setPassword()` method?

**Password Policies**

We discussed that it would be an option to implement support for password policies in user_ldap. Right now, I'm wondering what can be done to implement the detection of an expired password during login in user_ldap. It is not as easy as just parsing the bind result in `checkPassword()`, as the PHP LDAP module is not able to parse password policy related results. Instead, it would be required to check password policy related user attributes during login. For example, when using OpenLDAP, `checkPassword()` (or rather `processAttributes()`)would check "pwdChangedTime" and "pwdMaxAge" (more details [here](http://www.zytrax.com/books/ldap/ch6/ppolicy.html)). Would it be ok to implement this in user_ldap (as generic as possible), say in expert settings we could choose to check for password expiry and define the attribute names that are equivalent to the following:
- pwdPolicySubentry
- pwdChangedTime
- pwdMaxAge
- pwdExpireWarning
- pwdGraceUseTime
- pwdGraceAuthNLimit

Of course, just offering a checkbox to indicate password expiry checking and hardcoding the attributes would be fine for me, too :smile:

If the password expired, the user would have to be redirected to another page to change his password, not so sure how this could be done, maybe there is another module doing something similar? Or at least a way exists to popup a dialog with a link pointing to the change password page?

Your opinion is highly appreciated :smiley:

### CODE REVIEW

**Comment by GitHubUser4234** [comment]
_Reply from @blizzz (copied over from #23992):_

So essentially we would make the user (maybe optionally with a config) able to change his own password in our default LDAP backend. Thinking about this, I can imagine this happening, but it also has some edges:

- the configured LDAP user should have only the abilities to read, alone as a hardening measure (whether admins really pay attention, I do not know). Often users are allowed to reset their own passwords in LDAP. Going with this proper solution, we'd also need to specify the old password, so we can perform a bind. Either we need to extent the change password mechanism to be able to provide more input fields (more deep deep changes), or indeed we require the agent to be able to write passwords.

Given that password changes are coupled to one user instance, which is coupled to the backend instance, this if for know the only straightforward way. ( @ChristophWurst or did anything change in this regard)?

I'd put the checkbox into the Directory Settings (in the Advanced tab), unless you have better suggestions?

**Comment by GitHubUser4234** [comment]
Follow-up reply from @blizzz (copied over from #23992):

About the password policy:

I would prefer if we get the first things done and finished here. Usually it's best to keep changes and pull requests as small as possible and it seems for me that this is a good candidate to do as next step, in a different PR (btw., this also about extending user_ldap on password modification possibilities).

Now, is it necessary to have a setting here at all? For me it sounds pretty much like a thing that can be of interest anywhere. The attributes of interest can be read all read and inspected. Here it would be interesting to see whether this also works with AD or can be adopted to work with it.

If password changes are possible per user, the LDAP backend can redirect to an extra page of it's own which would end up in that reset form. In AppFramework days you would register a route and create a controller, I think it should be possible to mix it with the ancient way that the LDAP backend was written in. Does it help?

File: apps/user_ldap/lib/Access.php [hunk 1]
```diff
@@ -75,13 +76,19 @@ class Access extends LDAPUtility implements IUserTools {
 	* @var AbstractMapping $userMapper
 	*/
 	protected $groupMapper;
+	
+	/**
+	 * @var \OCA\user_ldap\lib\Helper
```
Comment (nickvergessen) [inline apps/user_ldap/lib/Access.php]: The correct class name is `\OCA\User_LDAP\Helper`

File: apps/user_ldap/appinfo/update.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+$helper = new \OCA\user_ldap\lib\Helper();
```
Comment (nickvergessen) [inline apps/user_ldap/appinfo/update.php]: The correct class name is `\OCA\User_LDAP\Helper`

File: apps/user_ldap/appinfo/install.php [hunk 3]
```diff
@@ -23,3 +24,6 @@
 if($state === 'doSet') {
 	OCP\Config::setSystemValue('ldapIgnoreNamingRules', false);
 }
+
+$helper = new \OCA\user_ldap\lib\Helper();
```
Comment (nickvergessen) [inline apps/user_ldap/appinfo/install.php]: The correct class name is `\OCA\User_LDAP\Helper`

File: apps/user_ldap/lib/Helper.php [hunk 4]
```diff
@@ -182,6 +183,70 @@ public function getDomainFromURL($url) {
...
+	
+	/**
+	 *
+	 * Set the LDAPProvider in the config
+	 *
+	 */
+	public function setLDAPProvider() {
+		$current = \OC::$server->getConfig()->getSystemValue('ldapProviderFactory', null);
+		if(is_null($current)) {
+			\OC::$server->getConfig()->setSystemValue('ldapProviderFactory', '\\OCA\\user_ldap\\lib\\LDAPProviderFactory');
```
Comment (nickvergessen) [inline apps/user_ldap/lib/Helper.php]: The correct class name is `\OCA\User_LDAP\LDAPProviderFactory`

File: apps/user_ldap/lib/iuserldap.php [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCA\user_ldap\lib;
```
Comment (nickvergessen) [inline apps/user_ldap/lib/iuserldap.php]: The correct namespace is `OCA\User_LDAP`

File: apps/user_ldap/lib/iuserldap.php [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+interface IUserLDAP {
```
Comment (nickvergessen) [inline apps/user_ldap/lib/iuserldap.php]: File name should be `apps/user_ldap/lib/IUserLDAP.php`

File: apps/user_ldap/lib/ldapprovider.php [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCA\user_ldap\lib;
```
Comment (nickvergessen) [inline apps/user_ldap/lib/ldapprovider.php]: The correct namespace is `OCA\User_LDAP`

File: apps/user_ldap/lib/ldapprovider.php [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+use OCP\LDAP\IDeletionFlagSupport;
+use OCP\IServerContainer;
+use OCA\user_ldap\USER_LDAP;
+use OCA\user_ldap\lib\User\DeletedUsersIndex;
+use OCA\user_ldap\mapping\UserMapping;
+
+/**
+ * LDAP provider for pulic access to the LDAP backend.
+ */
+class LDAPProvider implements ILDAPProvider, IDeletionFlagSupport {
```
Comment (nickvergessen) [inline apps/user_ldap/lib/ldapprovider.php]: File name should be `apps/user_ldap/lib/LDAPProvider.php`

File: apps/user_ldap/lib/ldapprovider.php [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+ *
+ */
+
+namespace OCA\user_ldap\lib;
+
+use OCP\IUserBackend;
+use OCP\LDAP\ILDAPProvider;
+use OCP\LDAP\IDeletionFlagSupport;
+use OCP\IServerContainer;
+use OCA\user_ldap\USER_LDAP;
```
Comment (nickvergessen) [inline apps/user_ldap/lib/ldapprovider.php]: The correct class name is `OCA\User_LDAP\User_LDAP`

File: apps/user_ldap/lib/ldapprovider.php [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+use OCA\user_ldap\lib\User\DeletedUsersIndex;
```
Comment (nickvergessen) [inline apps/user_ldap/lib/ldapprovider.php]: The correct class name is `OCA\User_LDAP\User\DeletedUsersIndex`

File: apps/user_ldap/lib/ldapprovider.php [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+use OCA\user_ldap\mapping\UserMapping;
```
Comment (nickvergessen) [inline apps/user_ldap/lib/ldapprovider.php]: The correct class name is `OCA\User_LDAP\Mapping\UserMapping`

File: apps/user_ldap/lib/ldapproviderfactory.php [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+namespace OCA\user_ldap\lib;
```
Comment (nickvergessen) [inline apps/user_ldap/lib/ldapproviderfactory.php]: The correct namespace is `OCA\User_LDAP`

File: apps/user_ldap/lib/ldapproviderfactory.php [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+use OCP\LDAP\ILDAPProviderFactory;
+
+class LDAPProviderFactory implements ILDAPProviderFactory {
```
Comment (nickvergessen) [inline apps/user_ldap/lib/ldapproviderfactory.php]: The correct file name is `apps/user_ldap/lib/LDAPProviderFactory.php
`

File: lib/public/ldap/ideletionflagsupport.php [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace OCP\LDAP;
+
+/**
+ * Interface IDeletionFlagSupport
+ *
+ * @package OCP\LDAP
+ * @since 9.1.0
+ */
+interface IDeletionFlagSupport {
```
Comment (nickvergessen) [inline lib/public/ldap/ideletionflagsupport.php]: File name needs to be `lib/public/LDAP/IDeletionFlagSupport.php`

File: lib/public/ldap/ildapprovider.php [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+namespace OCP\LDAP;
+
+/**
+ * Interface ILDAPProvider
+ *
+ * @package OCP\LDAP
+ * @since 9.1.0
+ */
+interface ILDAPProvider {
```
Comment (nickvergessen) [inline lib/public/ldap/ildapprovider.php]: File name needs to be `lib/public/LDAP/ILDAPProvider.php`

File: lib/public/ldap/ildapproviderfactory.php [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+/**
+ * Interface ILDAPProviderFactory
+ *
+ * This class is responsible for instantiating and returning an ILDAPProvider
+ * instance.
+ *
+ * @package OCP\LDAP
+ * @since 9.1.0
+ */
+interface ILDAPProviderFactory {
```
Comment (nickvergessen) [inline lib/public/ldap/ildapproviderfactory.php]: File name needs to be `lib/public/LDAP/ILDAPProviderFactory.php`

**Comment by GitHubUser4234** [comment]
Thanks a lot @nickvergessen for the comments. I applied the fixes to the master (PR #23992) and also updated this branch accordingly.

**Comment by DeepDiver1975** [comment]
I'm closing this because we have no signed cla and and future contribution of @GitHubUser4234 seem to flow into nextcloud. further contributions are still welcome anyhow. Please note that we continue the ldpa development in https://github.com/owncloud/user_ldap

**Pull request closed without merging**